compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/NameCollision.hs

```diff
@@ -107,41 +128,49 @@ fullyResolve = FRName . map T.toLower . \case
 -- | State of the name collision checker. This is a
 -- map from fully resolved names within a package to their
 -- original names. We update this map as we go along.
-newtype NCState = NCState (M.Map FRName [Name])
+data NCState = NCState (M.Map FRName [Name])
 
 -- | Initial name collision checker state.
 initialState :: NCState
 initialState = NCState M.empty
 
 -- | Monad in which to run the name collision check.
-type NCMonad t = S.StateT NCState (Either Error) t
+type NCMonad t = RWS () [Diagnostic] NCState t
 
 -- | Run the name collision with a blank initial state.
-runNameCollision :: NCMonad t -> Either Error t
-runNameCollision = flip S.evalStateT initialState
+runNameCollision :: NCMonad t -> [Diagnostic]
+runNameCollision m = snd (evalRWS m () initialState)
 
 -- | Try to add a name to the NCState. Returns Error only
 -- if the name results in a forbidden name collision.
-addName :: Name -> NCState -> Either Error NCState
-addName name (NCState nameMap) = do
-    let frName = fullyResolve name
-        oldNames = fromMaybe [] (M.lookup frName nameMap)
-        badNames = filter (nameCollisionForbidden name) oldNames
-    if null badNames then do
+addName :: Name -> NCState -> Either Diagnostic NCState
+addName name (NCState nameMap)
+    | null badNames =
         Right . NCState $ M.insert frName (name : oldNames) nameMap
-    else do
-        Left $ EForbiddenNameCollision
-            (displayName name)
-            (map displayName badNames)
+    | otherwise =
+        let err = EForbiddenNameCollision (displayName name) (map displayName badNames)
+            diag = toDiagnostic DsError err
+        -- If name is virtual are all badNames are virtual, we demote it to a
```

Comment:
According to the code, you probably mean:
```suggestion
        -- If name is virtual or all badNames are virtual, we demote it to a
```

Reply:
english is hard, thanks :)